src/platform/Zephyr/BLEAdvertisingArbiter.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "BLEAdvertisingArbiter.h"
+
+#include <lib/support/CodeUtils.h>
+#include <system/SystemError.h>
+
+namespace chip {
+namespace DeviceLayer {
+namespace {
+
+const BLEAdvertisingArbiter::Request & ToRequest(const sys_snode_t * node)
+{
+    return *static_cast<const BLEAdvertisingArbiter::Request *>(node);
+}
+
+} // namespace
+
+CHIP_ERROR BLEAdvertisingArbiter::RequestAdvertising(Request & request)
+{
+    sys_snode_t * prev = nullptr;
+    sys_snode_t * node;
+
+    // Find position of the new request in the list that preserves ordering by priority.
+    SYS_SLIST_FOR_EACH_NODE(&mRequests, node)
+    {
+        if (node == &request)
+        {
+            goto list_updated;
+        }
+
+        if (request.priority < ToRequest(node).priority)
+        {
+            break;
+        }
+
+        prev = node;
+    }
+
+    if (prev == nullptr)
+    {
+        sys_slist_prepend(&mRequests, &request);
+    }
+    else
+    {
+        sys_slist_insert(&mRequests, prev, &request);
+    }
+
+list_updated:
+    // If the new request is now top priority, restart advertising.
+    if (sys_slist_peek_head(&mRequests) == &request)
+    {
+        return Restart();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEAdvertisingArbiter::CancelAdvertising(Request & request)
+{
+    const sys_snode_t * head = sys_slist_peek_head(&mRequests);
+
+    VerifyOrReturnError(sys_slist_find_and_remove(&mRequests, &request), CHIP_ERROR_NOT_FOUND);
+
+    // If the cancelled request was top priority, restart advertising.
+    if (head == &request)
+    {
+        return Restart();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEAdvertisingArbiter::Restart()
+{
+    ReturnErrorOnFailure(System::MapErrorZephyr(bt_le_adv_stop()));
```

Comment:
Does it also fail when the advertising has been already stopped? Not sure if this is a real scenario but if so, the following code which starts a new advertising should get executed in such a case.

Reply:
Good point, I need to check not to fail in such a scenario.

Reply:
Seems that it returns no error in such a case. Added a comment about that.